Settings Search

luah5 opened the pull request:
### Description

This PR adds support for searching settings in CE's settings, it works very similar to the way System Settings searches.

### Related Issues

* close #1309

### Checklist

<!--- Add things that are not yet implemented above -->

- [x] I read and understood the [contributing guide](https://github.com/CodeEditApp/CodeEdit/blob/main/CONTRIBUTING.md) as well as the [code of conduct](https://github.com/CodeEditApp/CodeEdit/blob/main/CODE_OF_CONDUCT.md)
- [x] The issues this PR addresses are related to each other
- [x] My changes generate no new warnings
- [x] My code builds and runs on my machine
- [x] My changes are all related to the related issue above
- [x] I documented my code
- [x] Code is clean
- [x] Highlighted text of search
- [x] Fix grouping issue

### Screenshots

https://github.com/CodeEditApp/CodeEdit/assets/128280019/118ee588-9edb-4b73-a7f6-f70aa6f2d5e6

### Extra notes
I've updated the docs for adding a new setting so that it appears in the search results

austincondiff commented:
<img width="827" alt="image" src="https://github.com/CodeEditApp/CodeEdit/assets/806104/0caee033-043a-46c9-b4ca-2eb1c65a80f0">

You'll notice that:
1. The text is aligned with the parent item's label.
2. The child item itself is selectable.
3. There is only one parent item, in yours there is a parent for every child

luah5 commented:
> @luah5 :
> 
>  Registrazione.schermo.2023-06-13.alle.12.55.26.mov 
> 1. It would be possible for everyone in the same place to group them.
> 2. When clicked, on the text it should open the place where it is which happens, but you don't understand where that text is, it would be possible to make it flash or something similar to how android does when you search for something in the search bar settings.

Unfortunately, @Angelk90, it would be near-impossible to do number 2, as the way my code is structured with getting settings uses the `Models` not the `Views`.

austincondiff commented:
@Angelk90 I'd agree with out that it should probably flash or jump out at you in some way. System settings doesn't do this though so we might save this for another PR. Regardless, my guess is It is selecting all the related items because they all point to the same location, so they are all selected. We need to think of a workaround to this so only the item that is clicked is selected.

luah5 commented:
@Angelk90 I'm very aware of "grouping" issue I have already noted this down in this PR's description.

Wouter01 commented on CodeEdit/Features/Settings/SettingsView.swift:
```diff
@@ -7,47 +7,128 @@
+
+    /// Creates all the neccessary pages
+    private func populatePages() -> [SettingsPage] {
+        var pages: [SettingsPage] = []
+        let settingsData: SettingsData = SettingsData()
+
+        pages = createPageAndSettings(
+            settingsData.general,
+            parent: SettingsPage(.general, baseColor: .gray, icon: .system("gear")),
+            prePages: pages
+        )
+        pages = createPageAndSettings(
+            settingsData.accounts,
+            parent: SettingsPage(.accounts, baseColor: .blue, icon: .system("at")),
+            prePages: pages
+        )
+        pages = createPageAndSettings(
+            settingsData.theme,
+            parent: SettingsPage(.theme, baseColor: .pink, icon: .system("paintbrush.fill")),
+            prePages: pages
+        )
+        pages = createPageAndSettings(
+            settingsData.textEditing,
+            parent: SettingsPage(.textEditing, baseColor: .blue, icon: .system("pencil.line")),
+            prePages: pages
+        )
+        pages = createPageAndSettings(
+            settingsData.terminal,
+            parent: SettingsPage(.terminal, baseColor: .blue, icon: .system("terminal.fill")),
+            prePages: pages
+        )
+        pages = createPageAndSettings(
+            settingsData.sourceControl,
+            parent: SettingsPage(.sourceControl, baseColor: .blue, icon: .symbol("vault")),
+            prePages: pages
+        )
+        pages.append(SettingsPage(.location, baseColor: .green, icon: .system("externaldrive.fill")))
+
+        return pages
+    }
```
This is a very confusing pattern to build your pages. Consider letting the `createPageAndSettings` function return an array and combine all arrays afterwards

luah5 replied:
How would that look like in code?

Wouter01 reviewed (changes requested):
Nice work! I noticed some issues:
- Tab selection state isn't kept
- I should be able to go through the search results by pressing the up and down arrow keys
- It is difficult to search for settings because the search keywords don't match the "real" labels of the settings

This is a good base, but I think it needs improvements. I strongly suggest using settings labels instead of variable names as keywords. This is what will need to happen anyway.
Not doing that now is a bit pointless, as the work from this PR would be removed later on. I would also like to avoid complaints from users about search not being accurate / predictable

austincondiff commented:
@luah5 When there is no search text input, the menu item color needs to be primary. It looks like it is defaulting to secondary. It should only be secondary when there is text input into the search field.

luah5 commented:
So, I'm putting this PR back into Draft because as highlighted by @Wouter01's review comments something has to happen with naming. I'll post more in the discord, regarding what progress I'm making.

luah5 commented:
> Nice work! I noticed some issues:
> 
> * Tab selection state isn't kept

Could you show a video, does the tab selection need to be kept across sessions with `UserDefaults`?

> * I should be able to go through the search results by pressing the up and down arrow keys

Re implementing that feature makes the search results group again

> * It is difficult to search for settings because the search keywords don't match the "real" labels of the settings

Fixed.

> This is a good base, but I think it needs improvements. I strongly suggest using settings labels instead of variable names as keywords. This is what will need to happen anyway. Not doing that now is a bit pointless, as the work from this PR would be removed later on. I would also like to avoid complaints from users about search not being accurate / predictable

Fixed.

matthijseikelenboom commented on CodeEdit/Features/Settings/Models/SettingsPage.swift:
The file structure here is not in order. Enum should be defined above variables and variables should be defined above the init functions

thecoolwinter commented on CodeEdit/Features/Settings/Models/ModelNameToSettingName.swift:
We already have a `.strings` file in the Localization folder. These should be added there. Also I get a crash immediately when opening the settings view due to missing keys. 

A suggestion, instead of a standalone struct like this, a protocol for pages to conform to to be searchable, and conform all the settings pages to that protocol
```swift
protocol SettingsDataPage {
    static var searchKeys: Set<String> { get }
}

struct TextEditingSettings: Codable, Hashable, SettingsDataPage {
    // ...
    
    static var searchKeys: Set<String> {
        [
            "/* Put keys here to search */",
        ]
    }
}
```

This would allow you to remove all the mirroring stuff you've had to to (although I admit it's a cool way to do it).

thecoolwinter commented on CodeEdit/Features/Settings/SettingsView.swift:
```diff
@@ -7,50 +7,123 @@
+        return pages
+    }
+
+    /// Creates all the neccessary pages
```
`necessary` :)

thecoolwinter commented on CodeEdit/Features/Settings/SettingsView.swift:
```diff
@@ -7,50 +7,123 @@
+    }
+
     var body: some View {
+        let pages: [SettingsPage] = populatePages()
```
This is re-generated every view update, which is wasting resources. Lets keep the static property of pages if possible.

luah5 replied:
Fixed

thecoolwinter commented on CodeEdit/Features/Settings/SettingsView.swift:
```diff
@@ -7,50 +7,123 @@
-                    ForEach(Self.pages) { item in
-                        if searchText.isEmpty || item.name.rawValue.lowercased().contains(searchText.lowercased()) {
-                            SettingsPageView(item)
+                    ForEach(pages) { page in
```
If you use the `searchKeys` method suggested above, this could be simplified. Just use the old `ForEach` when the search is empty. And use an array of `SettingsSearchResult`s to represent the search results like so:
```swift
func searchResults() -> [SettingsSearchResult] {
    // Loop through each setting page, filtering each searchKey.
}
```
You'll have to modify `SettingsSearchResult` to contain a settings page and an array of strings to represent the search matches.

Then the view would just be a ForEach over each result with another ForEach inside that for all the matching keys.
```swift
if searchString.isEmpty {
    ForEach(Self.pages) // ...
} else {
    let results = searchResults()
    ForEach(results) { (page, keys) in
        ForEach(keys) // ...
    }
}
```

thecoolwinter commented on CodeEdit/Features/Settings/Models/SettingsPage.swift:
```diff
@@ -8,31 +8,31 @@
-        children: [SettingsPage] = []
+        baseColor: Color,
+        icon: IconResource,
+        isSetting: Bool = false,
```
I don't like using this struct for both search results and settings pages, it'll get confusing. I think remove the `isSetting` property here and don't use this for search results.

luah5 replied:
Unfortunately, that's not possible to change as the search results will group again, I have already tried to do so and it has taken an absurd amount of time. Sorry :(

thecoolwinter commented on CodeEdit/Features/Settings/Models/ModelNameToSettingName.swift:
```diff
@@ -0,0 +1,80 @@
+import Foundation
+import AppKit
+
+struct ModelNameToSettingName {
```
I get a crash here when opening, but I think this might be unnecessary. Instead of defining these here, define them on each page struct. That'll allow each struct to supply the localized string from the model instead of a standalone struct. You could define a protocol and make it available on a static property like:

```swift
protocol SearchableSettingsPage {
    static var searchKeys: [String] { get }
}

struct TextEditingSettings: Codable, Hashable, SearchableSettingsPage {
    // ...

    static var searchKeys: [String] {
        [
            "/* Search keys here */"
        ].map { NSLocalizedString($0, comment: "") }
    }
}
```

That'll be much easier to keep track of when updating/adding settings in the future.

thecoolwinter commented on CodeEdit/Features/Settings/Pages/LocationsSettings/Models/LocationsSettings.swift:
```diff
@@ -0,0 +1,52 @@
+        init() {}
+
+        /// Explicit decoder init for setting default values when key is not present in `JSON`
+        init(from decoder: Decoder) throws {
```
Storing this in the `settings.json` file makes it seem like we're supporting customizing the settings/themes/extensions folders when that's not the case. It's outside the scope of this PR. This struct should be removed or made into an `enum` with static properties for the URLs defined here.

thecoolwinter commented on CodeEdit/Utils/Extensions/String/String+highlightOccurrences.swift:
```diff
@@ -0,0 +1,51 @@
+import SwiftUI
+
+extension String {
+    func highlightOccurrences(_ ofSearch: String) -> some View {
```
Add some documentation here.

thecoolwinter commented on CodeEdit/Utils/Extensions/String/String+highlightOccurrences.swift:
Use a capitalized `HighlightOccurrences` name for this file to follow naming convention in the extensions folder.

thecoolwinter reviewed (changes requested):
This looks okay, however I think the changes to the settings page model needs to be reverted, as the model for searching can be simplified. Also see my comment for how pages should be able to define their own keys, instead of using mirroring. 

Both those changes should also simplify both the search itself and the UI to present the search.

Finally, the `LocationsSettings` struct should be removed. See my comment about it.

This is getting very close to finishing, you've made some great progress here and I think with just some small model/UI structure changes to make this more durable and less confusing it's good to merge.

austincondiff commented:
@luah5 Several things:
1. The text size is too small. Please use the original text size.
2. The default color of main navigation items is too light.
3. Result items text should line up with main items text. Give it a little more left spacing.

luah5 commented:
> @luah5 Several things:
> 
> 1. The text size is too small. Please use the original text size.

Fixed, I'm not sure what happened though.

> 2. The default color of main navigation items is too light.

<img width="70" alt="CodeEdit settings" src="https://github.com/CodeEditApp/CodeEdit/assets/128280019/2bc67b7e-932f-4dc9-afe0-d6de347462a4">

New: Right, Old: Left

I can't see a difference, but if you were referring to the text, I have fixed that.

> 3. Result items text should line up with main items text. Give it a little more left spacing.

Fixed in the latest commit

<img width="108" alt="CodeEdit settings" src="https://github.com/CodeEditApp/CodeEdit/assets/128280019/0789ca12-6258-4cb6-b9e3-a7940a5c8cb0">

austincondiff commented:
Font size and text color are still incorrect.

luah5 commented:
@Angelk90 Not reproducable, please check that you are on the latest commit.

austincondiff commented:
> @luah5 Several things:
> 
> 1. The text size is too small. Please use the original text size.
> 2. The default color of main navigation items is too light.
> 3. Result items text should line up with main items text. Give it a little more left spacing.

1. I just realized this is because of the sidebar size system setting. Disregard this.
2. This looks like it was addressed. Thanks!

luah5 commented:
> @luah5 :
> 
> Latest commit version, as you can see when I click on an element then it loses focus because the focus is on the input field.
> 
> Problem with font or size, looks weird.
> 
> mac os ventura.
> 
>  Registrazione.schermo.2023-07-10.alle.15.50.34.mov

Fixed :)

luah5 commented:
> @luah5 : Is it normal that it loses focus when I click?
> 
> https://github.com/CodeEditApp/CodeEdit/assets/20476002/81eae85e-7cb8-424b-b61f-fcee50311afd
> 
> 

That is not normal, and I cannot reproduce it, what version of macOS are you on?

luah5 commented:
I genuinely have no clue what's happening maybe a macOS bug. As you can see in this video the same issue doesn't occur on my machine, also check you are on the latest version. 

https://github.com/CodeEditApp/CodeEdit/assets/128280019/16821687-9fcb-44b1-a65e-37bbed60f4b3

Thanks for the report though, I appreciate it!

austincondiff commented:
Where are we on this PR? Is it ready to be approved? Is there anything pending revision?

luah5 commented:
@austincondiff, I'm just waiting for it to be reviewed.

matthijseikelenboom commented on CodeEdit/Features/Settings/Models/SettingsData.swift:
```diff
@@ -61,7 +62,42 @@ struct SettingsData: Codable, Hashable {
         self.featureFlags = try container.decodeIfPresent(FeatureFlagsSettings.self, forKey: .featureFlags) ?? .init()
     }
+
+    // swiftlint:disable cyclomatic_complexity
```
Cyclomatic complexity should be fixed. Is there no better way to this?

luah5 replied:
sorry, there isn't :c

matthijseikelenboom commented on CodeEdit/Features/Settings/Pages/SourceControlSettings/Models/SourceControlSettings.swift:
```diff
@@ -9,13 +9,39 @@ import Foundation
-    struct SourceControlSettings: Codable, Hashable {
+    struct SourceControlSettings: Codable, Hashable, SearchableSettingsPage {
+
+        var searchKeys: [String] {
+            [
+                "General",
+                "Enable source control",
+                "Refresh local status automatically",
+                "Fetch and refresh server status automatically",
+                "Add and remove files automatically",
+                "Select files to commit automatically",
+                "Show source control changes",
+                "Include upstream changes",
+                "Comparison view",
+                "Source control navigator",
+                "Default branch name",
+                "Git",
+                "Author Name",
+                "Author Email",
+                "Prefer to rebase when pulling",
+                "Show merge commits in per-file log"
+            ]
+            .map { NSLocalizedString($0, comment: "") }
+        }
```
Doesn't this fall apart when we support other languages?

luah5 replied:
Nope, because of the use of `NSLocalizedString`.

luah5 commented on CodeEdit/Features/Settings/Pages/GeneralSettings/Models/GeneralSettings.swift:
```diff
@@ -21,6 +21,28 @@ extension SettingsData {
+        var searchKeys: [String] {
+            [
+                "Appearance",
+                "File Icon Style",
```
Well, there's not really a way to read a SwiftUI View in swift.

austincondiff replied:
I'd tend to agree with this.

austincondiff replied:
Would it be possible to search the already existing labels by maybe generating this searchKeys array?

@Wouter01 seemed to have a pretty good suggestion by using a string table.

https://developer.apple.com/wwdc23/10155

austincondiff commented:
What is the status on this PR? We need to either finish it or close it. There is a lot of good work here so I'd rather push it over the finish line if we can.

luah5 commented:
> What is the status on this PR? We need to either finish it or close it. There is a lot of good work here so I'd rather push it over the finish line if we can.

Sorry for the recent inactivity, I'll try to get this finished with my limited time at the moment. Then this PR can be merged.

luah5 commented:
I would love to have this PR merged, could some people please review this PR?

matthijseikelenboom commented on CodeEdit/Features/Settings/Models/SettingsPage.swift:
```diff
@@ -59,4 +34,27 @@ struct SettingsPage: Hashable, Identifiable {
+    let icon: IconResource?
+
+    /// Default initializer
+    internal init(
```
`internal` keyword is not needed here

luah5 replied:
thanks for spotting that!

matthijseikelenboom reviewed (changes requested):
I've only that remark, other then that it looks fine

austincondiff commented on CodeEdit/Features/Settings/Pages/AccountsSettings/Models/AccountsSettings.swift:
```diff
@@ -10,10 +10,20 @@ import Foundation
+            [
+                "Accounts",
+                "Delete Account...",
+                "Add account..."
```
```suggestion
                "Add Account..."
```

0xWDG commented:
@luah5 can you fix the conflict with `CodeEdit.xcodeproj/project.pbxproj`?
Then it should be fine for re-review.

luah5 commented:
> @luah5 can you fix the conflict with `CodeEdit.xcodeproj/project.pbxproj`? Then it should be fine for re-review.

sure

luah5 commented:
just fixed the merge conflicts, if a reviewer has time, please review this PR, thanks!

0xWDG commented on CodeEdit/Features/Documents/Controllers/CodeEditWindowController.swift:
```diff
@@ -47,9 +47,7 @@ final class CodeEditWindowController: NSWindowController, NSToolbarDelegate, Obs
-    required init?(coder: NSCoder) {
-        fatalError("init(coder:) has not been implemented")
-    }
+    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }
```
```suggestion
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
```

0xWDG commented on CodeEdit/Features/Settings/Models/SettingsPage.swift:
```diff
@@ -59,4 +34,27 @@ struct SettingsPage: Hashable, Identifiable {
+    let baseColor: Color?
+    let isSetting: Bool
+    let settingName: String
+    var nameString: LocalizedStringKey { LocalizedStringKey(name.rawValue) }
```
```suggestion
    var nameString: LocalizedStringKey {
        LocalizedStringKey(name.rawValue)
    }
```

0xWDG commented on CodeEdit/Features/Settings/SettingsView.swift:
```diff
@@ -7,51 +7,84 @@
-
     /// Variables for the selected Page, the current search text and software updater
-    @State private var selectedPage = pages.first!
+    @State private var selectedPage: SettingsPage = .init(.general, baseColor: .gray, icon: .system("gear"))
```
IMHO it should be on multiple lines, it looks way more clear.

```suggestion
    @State private var selectedPage: SettingsPage = .init(
        .general, 
        baseColor: .gray, 
        icon: .system("gear")
    )
```

0xWDG commented on CodeEdit/Features/Settings/SettingsView.swift:
```diff
@@ -7,51 +7,84 @@
     var presentationMode
 
+    static var pages: [PageAndSettings] = [
+        .init(SettingsPage(.general, baseColor: .gray, icon: .system("gear"))),
+        .init(SettingsPage(.accounts, baseColor: .blue, icon: .system("at"))),
+        .init(SettingsPage(.theme, baseColor: .pink, icon: .system("paintbrush.fill"))),
+        .init(SettingsPage(.textEditing, baseColor: .blue, icon: .system("pencil.line"))),
+        .init(SettingsPage(.terminal, baseColor: .blue, icon: .system("terminal.fill"))),
+        .init(SettingsPage(.sourceControl, baseColor: .blue, icon: .symbol("vault"))),
+        .init(SettingsPage(.location, baseColor: .green, icon: .system("externaldrive.fill"))),
+        .init(SettingsPage(.featureFlags, baseColor: .cyan, icon: .system("flag.2.crossed.fill")))
```
I would suggest:

```suggestion
        .init(SettingsPage(
            .general, 
            baseColor: .gray, 
            icon: .system("gear")
        )),
        .init(SettingsPage(
            .accounts, 
            baseColor: .blue, 
            icon: .system("at")
        )),
        .init(SettingsPage(
            .theme, 
            baseColor: .pink, 
            icon: .system("paintbrush.fill")
        )),
        .init(SettingsPage(
            .textEditing,
            baseColor: .blue,
            icon: .system("pencil.line"))
        ),
        .init(SettingsPage(
            .terminal,
            baseColor: .blue,
            icon: .system("terminal.fill"))
        ),
        .init(SettingsPage(
            .sourceControl,
            baseColor: .blue,
            icon: .symbol("vault"))
        ),
        .init(SettingsPage(
            .location,
            baseColor: .green,
            icon: .system("externaldrive.fill"))
        ),
        .init(SettingsPage(
            .featureFlags,
            baseColor: .cyan,
            icon: .system("flag.2.crossed.fill"))
        )
```

0xWDG commented on CodeEdit/Utils/Extensions/String/String+HighlightOccurrences.swift:
```diff
@@ -0,0 +1,56 @@
+            let nonHighlightedText = self[currentIndex..<range.lowerBound]
+            let highlightedSubstring = self[range]
+
+            // swiftlint:disable shorthand_operator
```
Isn't there a better solution? for merging the text?

0xWDG replied:
Couldn't find one.
got another small point.

0xWDG reviewed (changes requested):
I found a few styling 'issues'.

0xWDG commented on CodeEdit/Utils/Extensions/String/String+HighlightOccurrences.swift:
```diff
@@ -0,0 +1,56 @@
+
+            // swiftlint:disable shorthand_operator
+            highlightedText = highlightedText + Text(nonHighlightedText).foregroundColor(.secondary)
+            highlightedText = highlightedText + Text(highlightedSubstring).foregroundColor(.primary)
```
I would suggest to make it bold.
on my machine the text color change make no clear visual difference on screen.

```suggestion
            highlightedText = highlightedText + Text(highlightedSubstring).foregroundColor(.primary).bold()
```

<table><tr><th>Without Bold</th><th>With <strong>bold()</strong></th></tr>
<td>
<img width="220" alt="image" src="https://github.com/CodeEditApp/CodeEdit/assets/1290461/df50ff8d-f29f-4286-a655-56c1ddfa37c3">

</td><td>
<img width="209" alt="image" src="https://github.com/CodeEditApp/CodeEdit/assets/1290461/ae5bef4f-94a7-4152-bae9-1f551f7d2b24">

</td></tr></table>

0xWDG commented:
@luah5 FYI, `shorthand_operator` should not trigger if you use `Text() + Text()`.
see https://github.com/realm/SwiftLint/issues/4936, https://github.com/realm/SwiftLint/issues/4953 for more information.

austincondiff commented:
In the interest of preserving the work that has been done and not letting this PR become stale, I merged this PR. For those that requested changes to this PR, feel free to open follow-up PR's for these changes.
